src/ast/analysis/TypeConstraints.cpp

```diff
@@ -484,56 +486,75 @@ void TypeConstraintsAnalysis::visit_(type_identity<BinaryConstraint>, const Bina
-                addConstraint(subtypesOfTheSameBaseType(var, functorVar));
+    auto argVars = map(fun.getArguments(), [&](auto&& x) { return getVar(x); });
+    // The type of the user-defined function might not be set at this stage.
+    // If so then add overloads as alternatives
+    if (!typeAnalysis.hasValidTypeInfo(fun))
+        addConstraint(satisfiesOverload(typeEnv, functorBuiltIn(fun.getBaseFunctionOp()), functorVar, argVars,
+                isInfixFunctorOp(fun.getBaseFunctionOp())));
+
+    // In polymorphic case
+    // We only require arguments to share a base type with a return type.
+    // (instead of, for example, requiring them to be of the same type)
+    // This approach is related to old type semantics
+    // See #1296 and tests/semantic/type_system4
+    if (isInfixFunctorOp(fun.getBaseFunctionOp())) {
+        for (auto&& var : argVars)
+            addConstraint(subtypesOfTheSameBaseType(var, functorVar));
 
-            return;
-        }
-
-        if (!typeAnalysis.hasValidTypeInfo(intrFun)) {
-            return;
-        }
+        return;
     }
 
-    // Skip constraint adding if type info is not available
-    if (!typeAnalysis.hasValidTypeInfo(&fun)) {
+    if (!typeAnalysis.hasValidTypeInfo(fun)) {
         return;
     }
 
     // add a constraint for the return type of the functor
-    TypeAttribute returnType = typeAnalysis.getFunctorReturnType(&fun);
+    TypeAttribute returnType = typeAnalysis.getFunctorReturnTypeAttribute(fun);
     addConstraint(isSubtypeOf(functorVar, typeEnv.getConstantType(returnType)));
     // Special case. Ord returns the ram representation of any object.
-    if (intrFun && typeAnalysis.getPolymorphicOperator(intrFun) == FunctorOp::ORD) {
+    if (typeAnalysis.getPolymorphicOperator(fun) == FunctorOp::ORD) {
         return;
     }
 
     // Add constraints on arguments
     auto arguments = fun.getArguments();
     for (size_t i = 0; i < arguments.size(); ++i) {
-        TypeAttribute argType = typeAnalysis.getFunctorArgType(&fun, i);
+        TypeAttribute argType = typeAnalysis.getFunctorParamTypeAttribute(fun, i);
         addConstraint(isSubtypeOf(getVar(arguments[i]), typeEnv.getConstantType(argType)));
     }
 }
 
+void TypeConstraintsAnalysis::visit_(type_identity<UserDefinedFunctor>, const UserDefinedFunctor& fun) {
+    auto functorVar = getVar(fun);
+
+    // I found this very confusing, hopefully this comment helps someone else.
+    // I assumed that this branch cannot be taken, because the Semantic checker
+    // verifies that every functor has a declaration!  However, it turns out that
+    // the SemanticChecker is *not* the first Transformer which gets run and so
+    // it's not really clear that those invariants hold yet!
+    // I don't particularly like that but am not at a place where I can change the
+    // order of the passes/transformers.  So, for now, here's a comment for the next
+    // person going doing this rabbit hole.
```

Comment:
This is right! The pipeline is quite complex. For example components are instantiated first and the type analysis is required by the semantic checker. It is not clear for the analysis whether the program is semantically correct at this point.

Reply:
I think it would be nice, at some point, to "formalize" these requirements.  Perhaps there could be a "normalization" part of the pipeline which can't rely on certain things existing/being correct, and a "post-normalization", where these invariants can be expected to hold.
While this would be nice from a "software perspective", I am not sure what the overall value to the project would be.  Clearly there are higher-priority concerns.